integration/metrics_test.go

```diff
@@ -192,3 +201,204 @@ disable_verification = %v
+		})
+	}
+}
+
+// buildIndexByManipulatingZtocData produces a new soci index by manipulating
+// the ztocs of an existing index specified by `indexDigest`.
+//
+// The new index (and ztocs) are stored separately and the original index keeps unchanged.
+// The manipulated ztocs are (de)serializable but have meaningless ztoc data (manipuated by `manipulator`).
+// This helps test soci behaviors when ztocs have valid format but wrong/corrupted data.
+func buildIndexByManipulatingZtocData(sh *shell.Shell, indexDigest string, manipulator func(*ztoc.Ztoc)) (string, error) {
+	index, err := sociIndexFromDigest(sh, indexDigest)
+	if err != nil {
+		return "", err
+	}
+
+	var ztocDescs []ocispec.Descriptor
+	for _, blob := range index.Blobs {
+		ztocDigest := blob.Digest.String()
+		blobContent := fetchContentFromPath(sh, blobStorePath+"/"+trimSha256Prefix(ztocDigest))
+		zt, err := ztoc.Unmarshal(bytes.NewReader(blobContent))
+		if err != nil {
+			return "", fmt.Errorf("invalid ztoc %s from soci index %s: %v", ztocDigest, indexDigest, err)
+		}
+
+		// manipulate the ztoc
+		manipulator(zt)
+
+		ztocReader, ztocDesc, err := ztoc.Marshal(zt)
+		if err != nil {
+			return "", fmt.Errorf("unable to marshal ztoc %s: %s", ztocDesc.Digest.String(), err)
+		}
+		ztocBytes, err := io.ReadAll(ztocReader)
+		if err != nil {
+			return "", fmt.Errorf("unable to read bytes of ztoc %s: %s", ztocDesc.Digest.String(), err)
+		}
+
+		ztocPath := fmt.Sprintf("%s/%s", blobStorePath, trimSha256Prefix(ztocDesc.Digest.String()))
+		if err := testutil.WriteFileContents(sh, ztocPath, ztocBytes, 0600); err != nil {
+			return "", fmt.Errorf("cannot write ztoc %s to path %s: %s", ztocDesc.Digest.String(), ztocPath, err)
+		}
+
+		ztocDesc.MediaType = soci.SociLayerMediaType
+		ztocDesc.Annotations = blob.Annotations
+		ztocDescs = append(ztocDescs, ztocDesc)
+	}
+
+	newIndex := soci.Index{
+		MediaType:    soci.OCIArtifactManifestMediaType,
+		ArtifactType: soci.SociIndexArtifactType,
+		Blobs:        ztocDescs,
+		Subject: &ocispec.Descriptor{
+			MediaType: soci.OCIArtifactManifestMediaType,
+			Digest:    index.Subject.Digest,
+			Size:      index.Subject.Size,
+		},
+	}
+
+	b, err := json.Marshal(newIndex)
+	if err != nil {
+		return "", err
+	}
+
+	newIndexDigest := digest.FromBytes(b)
+	newIndexPath := fmt.Sprintf("%s/%s", blobStorePath, trimSha256Prefix(newIndexDigest.String()))
+	if err := testutil.WriteFileContents(sh, newIndexPath, b, 0600); err != nil {
+		return "", fmt.Errorf("cannot write index %s to path %s: %s", newIndexDigest, newIndexPath, err)
+	}
+	return strings.Trim(newIndexDigest.String(), "\n"), nil
+}
+
+// checkFuseOperationFailureMetrics checks if output from metrics endpoint includes
+// a specific fuse operation failure metrics (or any fuse op failure if an empty string is given)
+func checkFuseOperationFailureMetrics(t *testing.T, output string, metricToCheck string, expectOpFailure bool) {
```

Comment:
Can this also return an error instead of doing `t.Fatal`?

Reply:
yeah I thought about this too, but `t` is used to log fuse op failure lines. That'll be helpful to debug if we have unexpected fuse failures in the tests.